site/content/en/latest/user/private-key-provider.md

```diff
@@ -0,0 +1,306 @@
+                    operator: Exists
+                  - key: feature.node.kubernetes.io/cpu-cpuid.AVX512VBMI2
+                    operator: Exists
+EOF
+```
+
+Or using `preferredDuringSchedulingIgnoredDuringExecution` for best effort scheduling, or not doing any node affinity, just doing the random scheduling. The CryptoMB private key provider supports software fallback if the required CPU instructions aren't here.
+
+## Apply EnvoyPatchPolicy to enable private key provider
+
+### For QAT
+
+```shell
+cat <<EOF | kubectl apply -f -
+apiVersion: gateway.envoyproxy.io/v1alpha1
+kind: EnvoyPatchPolicy
+metadata:
+  name: key-provider-patch-policy
+  namespace: default
+spec:
+  targetRef:
+    group: gateway.networking.k8s.io
+    kind: Gateway
+    name: eg
+    namespace: default
+  type: JSONPatch
+  jsonPatches:
+    - type: "type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.Secret"
+      name: default-example-cert
+      operation:
+        op: add
+        path: "/tls_certificate/private_key_provider"
+        value:
+          provider_name: qat
+          typed_config:
+            "@type": "type.googleapis.com/envoy.extensions.private_key_providers.qat.v3alpha.QatPrivateKeyMethodConfig"
+            private_key:
+              inline_string: |
+                abcd
+            poll_delay: 0.001s
+          fallback: true
+    - type: "type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.Secret"
+      name: default-example-cert
+      operation:
+        op: copy
+        from: "/tls_certificate/private_key"
+        path: "/tls_certificate/private_key_provider/typed_config/private_key"
+EOF
+```
+
+### For CryptoMB
+
+```shell
+cat <<EOF | kubectl apply -f -
+apiVersion: gateway.envoyproxy.io/v1alpha1
+kind: EnvoyPatchPolicy
+metadata:
+  name: key-provider-patch-policy
+  namespace: default
+spec:
+  targetRef:
+    group: gateway.networking.k8s.io
+    kind: Gateway
+    name: eg
+    namespace: default
+  type: JSONPatch
+  jsonPatches:
+    - type: "type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.Secret"
+      name: default-example-cert
+      operation:
+        op: add
+        path: "/tls_certificate/private_key_provider"
+        value:
+          provider_name: cryptomb
+          typed_config:  
+            "@type": "type.googleapis.com/envoy.extensions.private_key_providers.cryptomb.v3alpha.CryptoMbPrivateKeyMethodConfig"
+            private_key:
+              inline_string: |
+                abcd
+            poll_delay: 0.001s
+          fallback: true
+    - type: "type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.Secret"
+      name: default-example-cert
+      operation:
+        op: copy
+        from: "/tls_certificate/private_key"
+        path: "/tls_certificate/private_key_provider/typed_config/private_key"
+EOF
+```
+
+## Testing
+
+Follow the instructions in [TLS Termination for TCP](./tls-termination) to do the test.
```

Comment:
can we create two listeners or gateway ? to test with and w/o key provider ? and highlight the differentiating factor ? e.g. time to complete the request in client or some way to show time taken to complete tls handshake in envoy

Reply:
ok, we should be able to do that.

Reply:
I added fortio command before enabling the envoy patch policy and after to benchmark the gateway simply. Then the people can compare the results to know the private key provider is working.